gobblin-core/src/test/java/gobblin/source/extractor/extract/AbstractSourceTest.java

```diff
@@ -0,0 +1,235 @@
+package gobblin.source.extractor.extract;
+
+import com.google.common.collect.Lists;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.configuration.SourceState;
+import gobblin.configuration.State;
+import gobblin.configuration.WorkUnitState;
+import gobblin.configuration.WorkUnitState.WorkingState;
+import gobblin.source.extractor.Extractor;
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+
+public class AbstractSourceTest {
+
+  WorkUnitState committedWorkUnitState;
```

Comment:
I suggest you add an additional member that is a list of work units having state FAILED or SUCCESSFUL. This makes it easier to do the assert later. You just need to compare this list with the list returned from getPreviousWorkUnitStatesForRetry.